### PR TITLE
Support merging StateUpdates

### PR DESCRIPTION
**Pull request by syuhei176**
## Overview

Update StateChecker to support merge transaction.

## Task List

- [x] Support merge state in child chain
- [x] Support merge state in wallet

Related API.
Search states which can be merged and send merge transaction.
 
```
wallet.merge()
```

related to #290 

### compatibility

* APIs aren't changes
* data structure of "utxos" was changed

### CODE REVIEW

File: packages/core/src/SegmentChecker.ts [hunk 1]
```diff
@@ -13,26 +12,38 @@ export class SegmentChecker {
...
     hash: string,
     stateUpdate: StateUpdate,
     deprecationWitness: string
   ) {
-    return this.leaves.filter(l => {
-      return l.verifyDeprecation(hash, stateUpdate, deprecationWitness, this.predicatesManager)
-    }).length > 0
+    const targets = this.getLeavesInRange(stateUpdate.getSegment())
+    return targets.filter(l => {
+      return l.verifyDeprecation(hash, stateUpdate.getSubStateUpdate(l.getSegment()), deprecationWitness, this.predicatesManager)
```
Comment (shogochiai) [inline packages/core/src/SegmentChecker.ts]: Must have resource to review here
https://github.com/cryptoeconomicslab/plasma-chamber/issues/223#issuecomment-478184668

File: packages/core/test/SegmentChecker.test.ts [hunk 2]
```diff
@@ -45,6 +45,12 @@ describe('SegmentChecker', () => {
     predicate,
     AliceAddress
   )
+  const merge = OwnershipPredicate.create(
+    Segment.ETH(ethers.utils.bigNumberify(0), ethers.utils.bigNumberify(15000000)),
+    ethers.utils.bigNumberify(8),
+    predicate,
```
Comment (shogochiai) [inline packages/core/test/SegmentChecker.test.ts]: `const predicate = AliceAddress`

...why?

Comment (syuhei176) [reply packages/core/test/SegmentChecker.test.ts]: in this test, any predicate address is ok.

File: packages/wallet/src/StateManager.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  constructor(predicatesManager: PredicatesManager) {
+    super(predicatesManager)
+  }
+  
+  isContain(tx: SignedTransactionWithProof): boolean {
+    return this._isContain(tx.getTxHash(), tx, tx.getTransactionWitness())
+  }
+
+  spend(tx: SignedTransactionWithProof) {
```
Comment (shogochiai) [inline packages/wallet/src/StateManager.ts]: spend is verb. It implies to modify something. But why is this returning bool?

Comment (syuhei176) [reply packages/wallet/src/StateManager.ts]: modification affected to this.leaves(insert from BaseStateManeger) and just return result.

File: packages/core/src/state/BaseStateManager.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    if(canDeprecate) {
+      this.leaves = this.leaves.filter(l => {
+        return targets.map(t => t.getStateHash()).indexOf(l.getStateHash()) < 0
+      })
+      targets.forEach(target => {
+        target.getRemainingState(stateUpdate).forEach(newTxo => {
+          this.leaves.push(newTxo)
+        })
+      })
+      return true
```
Comment (shogochiai) [inline packages/core/src/state/BaseStateManager.ts]: hmm, boolean return as a success report of mutation...

**Comment by syuhei176** [comment]
Add about compatibility in description.

* API aren't changes
* data structure of "utxos" was changed

File: packages/core/src/state/BaseStateManager.ts (hunk 4 above)
Comment (syuhei176) [reply packages/core/src/state/BaseStateManager.ts]: changed return type to spent states.
Now `spend` returns spent state.

File: packages/core/test/SegmentChecker.test.ts (hunk 2 above)
Comment (syuhei176) [reply packages/core/test/SegmentChecker.test.ts]: Changed predicate to `ethers.constants.AddressZero`.

File: packages/core/test/StateManager.test.ts [hunk 5]
```diff
@@ -11,7 +11,7 @@ describe('StateManager', () => {
   const AliceAddress = utils.computeAddress(AlicePrivateKey)
   const BobAddress = utils.computeAddress(BobPrivateKey)
 
-  const predicate = AliceAddress
+  const predicate = constants.AddressZero
```
Comment (shogochiai) [inline packages/core/test/StateManager.test.ts]: noted

File: packages/core/src/state/BaseStateManager.ts [hunk 6]
```diff
@@ -62,9 +76,9 @@ export class BaseStateManager {
           this.leaves.push(newTxo)
         })
       })
-      return true
+      return targets
```
Comment (shogochiai) [inline packages/core/src/state/BaseStateManager.ts]: noted

**Pull request merged**